### PR TITLE
feat(lsp): support for completions

### PR DESCRIPTION
**Pull request by baszalmstra**
Adds initial support for completions for IDEs. This is far from finished, there is a lot more to add but I didn't want to make this too big. Currently supports two types of completion:

* Completion of names in scope
* Completion of struct fields

I used the same design as Rust Analyzer. 
 
image

This PR depends on #298

### CODE REVIEW

File: crates/mun_hir/src/semantics.rs [hunk 1]
```diff
@@ -0,0 +1,226 @@
+//! `Semantics` provides the means to get semantic information from syntax trees that are not
+//! necessarily part of the compilation process. This is useful when you want to extract information
+//! from a modified source file in the context of the current state.
+//!
+//! Our compilation databases (e.g. `HirDatabase`) provides a lot of steps to go from a syntax tree
+//! (as provided by the [`mun_syntax::ast`] module) to more abstract representations of the source
+//! through the process of `lowering`. However, for IDE purposes we often want to cut through all
+//! this and go from source locations straight to lowered datastructures and back. This is what the
```
Comment (Wodann) [inline crates/mun_hir/src/semantics.rs:8]:
```suggestion
//! this and go from source locations straight to lowered data structures and back. This is what
```

File: crates/mun_hir/src/semantics.rs [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    /// Constructs a new `Semantics` instance with the given database.
+    pub fn new(db: &'db dyn HirDatabase) -> Self {
+        Self {
+            db,
+            source_file_to_file: Default::default(),
+            source_to_definition_cache: Default::default(),
+        }
+    }
+
+    /// Returns the the Concrete Syntax Tree for the file with the given `file_id`.
```
Comment (Wodann) [inline crates/mun_hir/src/semantics.rs:48]:
```suggestion
    /// Returns the Concrete Syntax Tree for the file with the given `file_id`.
```

File: crates/mun_hir/src/semantics.rs [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    fn with_source_to_def_context<F: FnOnce(&mut SourceToDefContext) -> T, T>(&self, f: F) -> T {
+        let mut cache = self.source_to_definition_cache.borrow_mut();
+        let mut context = SourceToDefContext {
+            db: self.db,
+            cache: &mut *cache,
+        };
+        f(&mut context)
+    }
+
+    /// Returns the file that is associated with the given root CST node or None if no such
```
Comment (Wodann) [inline crates/mun_hir/src/semantics.rs:114]:
```suggestion
    /// Returns the file that is associated with the given root CST node or `None` if no such
```

File: crates/mun_hir/src/semantics.rs [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+impl Local {
+    /// Returns the type of this local
+    pub fn ty(self, db: &dyn HirDatabase) -> Ty {
+        let infer = db.infer(self.parent);
+        infer[self.pat_id].clone()
+    }
+}
+
+impl<'a> SemanticsScope<'a> {
+    pub fn visit_all_names(&self, visitor: &mut dyn FnMut(Name, ScopeDef)) {
```
Comment (Wodann) [inline crates/mun_hir/src/semantics.rs:204]: Missing docs

File: crates/mun_hir/src/semantics/source_to_def.rs [hunk 5]
```diff
@@ -0,0 +1,154 @@
+use crate::code_model::src::HasSource;
```
Comment (Wodann) [inline crates/mun_hir/src/semantics/source_to_def.rs:1]: Merge use statements

File: crates/mun_hir/src/semantics/source_to_def.rs [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+use crate::ids::{DefWithBodyId, FunctionId, ItemDefinitionId, Lookup, StructId, TypeAliasId};
+use crate::item_scope::ItemScope;
+use crate::{DefDatabase, FileId, HirDatabase, InFile, ModuleId};
+use mun_syntax::{ast, match_ast, AstNode, SyntaxNode};
+use rustc_hash::FxHashMap;
+
+pub(super) type SourceToDefCache = FxHashMap<SourceToDefContainer, SourceToDefMap>;
+
+pub(super) struct SourceToDefContext<'a, 'b> {
```
Comment (Wodann) [inline crates/mun_hir/src/semantics/source_to_def.rs:10]: Missing docs

File: crates/mun_hir/src/source_analyzer.rs [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+use crate::ids::DefWithBodyId;
+use crate::{
+    resolver_for_scope, Body, ExprId, ExprScopes, FileId, HirDatabase, InFile, InferenceResult,
+    Resolver, Ty,
+};
+use mun_syntax::{ast, AstNode, SyntaxNode, TextRange, TextSize};
+use std::sync::Arc;
+
+/// A `SourceAnalyzer` is a wrapper which exposes the HIR API in terms of the original source file.
+/// Its useful to query things from the syntax.
```
Comment (Wodann) [inline crates/mun_hir/src/source_analyzer.rs:12]:
```suggestion
/// It's useful to query things from the syntax.
```

File: crates/mun_hir/src/source_analyzer.rs [hunk 8]
```diff
@@ -0,0 +1,171 @@
+use crate::expr::scope::LocalScopeId;
```
Comment (Wodann) [inline crates/mun_hir/src/source_analyzer.rs:1]: Merge use statements

File: crates/mun_language_server/src/completion.rs [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+/// This is the main entry point for computing completions. This is a two step process.
+///
+/// The first step is to determine the context of where the completion is requested. This
+/// information is captured in the [`CompletionContext`]. The context captures things like which
+/// type of syntax node is before the cursor or the current scope.
+///
+/// Second is to compute a set of completions based on the previously computed context. We provide
+/// several methods for computing completions based on different syntax contexts. For instance when
+/// writing `foo.$0` you want to complete the fields of `foo` and dont want the local variables of
```
Comment (Wodann) [inline crates/mun_language_server/src/completion.rs:31]:
```suggestion
/// writing `foo.$0` you want to complete the fields of `foo` and don't want the local variables of
```

File: crates/mun_language_server/src/completion/context.rs [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    pub fn new(db: &'a AnalysisDatabase, position: FilePosition) -> Option<Self> {
+        let sema = Semantics::new(db);
+
+        let original_file = sema.parse(position.file_id);
+
+        // Insert a fake identifier to get a valid parse tree. This tree will be used to determine
+        // context. The actual original_file will be used for completion.
+        let file_with_fake_ident = {
+            let parse = db.parse(position.file_id);
+            let edit = Indel::insert(position.offset, String::from("intellijRulezz"));
```
Comment (Wodann) [inline crates/mun_language_server/src/completion/context.rs:39]: Mun Rulezz!

File: crates/mun_language_server/src/completion/context.rs [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+            // ident, so it should have the same range in the non-modified file
+            self.dot_receiver = field_expr
+                .expr()
+                .map(|e| e.syntax().text_range())
+                .and_then(|r| find_node_with_range(original_file, r));
+        }
+    }
+}
+
+/// Returns true if the given `node` or on if its parents is of the specified type.
```
Comment (Wodann) [inline crates/mun_language_server/src/completion/context.rs:127]:
```suggestion
/// Returns true if the given `node` or one of its parents is of the specified type.
```

**Review by Wodann — CHANGES_REQUESTED** [review]
Generally looks good. It's a lot of complex code though, so I didn't comprehend it all. I noticed that quite a lot of code was untested. As this is (seemingly) complex code, it'd be good to add a broader suite of tests (as we keep expanding on our LSP.

File: crates/mun_language_server/src/completion/context.rs (hunk 10 above)
Comment (baszalmstra) [reply crates/mun_language_server/src/completion/context.rs:39]: "intellijRulezz" is a standard string used for this purpose

File: crates/mun_hir/src/code_model/struct.rs [hunk 12]
```diff
@@ -29,25 +29,34 @@ impl From<StructId> for Struct {
...
     }
 
-    pub fn id(self) -> LocalStructFieldId {
+    /// Returns the index of this field in the parent
+    pub fn index(self, _db: &dyn HirDatabase) -> u32 {
+        self.id.into_raw().into()
+    }
+
+    /// Returns the ID of the field with relation to the parent struct
+    pub(crate) fn id(self) -> LocalStructFieldId {
```
[the remark this answers is not in the document]
Comment (baszalmstra) [reply crates/mun_hir/src/code_model/struct.rs:59]: I renamed `LocalStructFieldId` to `LocalFieldId` and `StructFieldData` to `FieldData`

File: crates/mun_hir/src/semantics.rs (hunk 4 above)
Comment (baszalmstra) [reply crates/mun_hir/src/semantics.rs:205]: Added.

File: crates/mun_hir/src/semantics.rs [hunk 13]
```diff
@@ -201,6 +201,7 @@ impl Local {
 }
 
 impl<'a> SemanticsScope<'a> {
+    /// Call the `visit` function for every named item in the scope
```
Comment (Wodann) [inline crates/mun_hir/src/semantics.rs:204]: `visitor` function

**Review by Wodann — APPROVED** [review]

**Pull request merged**